toc.md

```diff
@@ -428,77 +429,232 @@ An additional backlog of 936 hours of image processing jobs everyday, along with
+![efficiency_algorithms_benchmark](/images/efficiency_algorithms_benchmark_1.png){:width="400"}
 
-*Note: we want to quickly mention that we did not build Workerholic to compete with Sidekiq, and that you should not prefer our library to theirs' or vice-versa. We just chose Sidekiq to benchmark against because it is the leader of background job processing in Ruby and we thought that is the bar we should aim for. Could we do better? Maybe, maybe not. We haven't tried to yet, because we were satisfied with the current results.*
+As shown in the above diagram, Sidekiq's random polling algorithm took 242 seconds. We decided to start with a simple and dynamic algorithm: evenly-balancing workers.
 
 ##### Evenly balanced workers
+
+The purpose of this algorithm is to attribute a fair amount of workers between each queue by dynamically assigning a queue to a worker, every second. This way if a queue is empty the workers that were polling from it can be evenly redistributed between the non-empty queues.
+
+Once we had a working implementation of the algorithm we decided to benchmark with the same scenario as we did for Sidekiq:
+
+![efficiency_algorithms_benchmark](/images/efficiency_algorithms_benchmark_2.png){:width="400"}
+
+The algorithm performed poorly compared to Sidekiq's random polling. It took 357 seconds to perform the same amount and type of jobs with our algorithm. This is because in the context of having CPU-blocking jobs, having multiple workers would not make a difference as compared to having one woker (when using MRI).
+
+How can Workerholic perform better?
+
 ##### Adaptive and Successive Algorithm (ASP)
+###### Iteration 1
+
+We decided to use a different, custom, algorithm: Adaptive and Successive Provisioning of workers based on queue load. This would allow workers to switch faster to IO queues in case there are any because CPU-bound are usually executed a faster rate than IO blocking jobs. Once again, we benchmarked using the same amount and type of jobs:
+
+![efficiency_algorithms_benchmark](/images/efficiency_algorithms_benchmark_3.png){:width="400"}
+
+The jobs were executed faster than our evenly-balancing algorithm but still much slower than Sidekiq's random polling. We thought we could do better and make Workerholic more efficient, which is why we went through a second iteration for our ASP algorithm.
+
+###### Iteration 2
+
+We decided that we needed a way to identify queues containing IO-bound from CPU-bound jobs. Our first iteration of ASP was an improvement, but in the case of having some queues containing CPU-bound jobs, we were still potentially assigning more than workers.
```

Comment:
We decided that we needed a way to distinguish queues containing IO-bound and CPU-bound jobs. Our first iteration of ASP was an improvement, but in the case of having queues containing CPU-bound jobs, we were still potentially assigning more than 1 worker.